runtime/dynstats.c

```diff
@@ -499,12 +643,161 @@ dynstats_createCtr(dynstats_bucket_t *b, const uchar* metric, dynstats_ctr_t **c
 	RETiRet;
 }
 
+/* not supported in older platforms - disable */
+#ifndef O_CLOEXEC
+#define O_CLOEXEC 0
+#endif
+/*
+ *	Helper function adapted from imfile of the same name - possible candidate for refactor.
+ */
+static rsRetVal ATTR_NONNULL()
+writeStateFile(dynstats_bucket_t *b, const char *file_name, const char *content) {
+	const size_t toWrite = strlen(content);
+	ssize_t w = 0;
+	DEFiRet;
+	const int fd = open(file_name, O_CLOEXEC | O_NOCTTY | O_WRONLY | O_CREAT | O_TRUNC, 0600);
+	if(fd < 0) {
+		LogError(errno, RS_RET_IO_ERROR, "dynstats: cannot open state file '%s' for "
+			"persisting file state - some data will probably be duplicated "
+			"on next startup", file_name);
+		ABORT_FINALIZE(RS_RET_IO_ERROR);
+	}
+
+	w = write(fd, content, toWrite);
+	if(w != (ssize_t) toWrite) {
+		LogError(errno, RS_RET_IO_ERROR, "dynstats: partial write to state file '%s' "
+			"this may cause trouble in the future. We will try to delete the "
+			"state file, as this provides most consistent state", file_name);
+		unlink(file_name);
+		ABORT_FINALIZE(RS_RET_IO_ERROR);
+	}
+
+finalize_it:
+	/* TODO: revisit - seems wasteful to use the bucket lock for this */
+	pthread_rwlock_wrlock(&b->lock);
+	if (iRet == RS_RET_IO_ERROR) {
+		b->flushedErrors++;
+	}
+	if (w != -1) {
+		b->flushedBytes += w;
+		b->flushedCount++;
+	}
+	pthread_rwlock_unlock(&b->lock);
```

Comment:
Looks like this can race with destroy, destroy actually kills the lock and bucket.

Reply:
yes, great catch. I've moved all instances of writers prior to destroy. added some asserts into this function to confirm as well.